handlers/proxy.go

```diff
@@ -1,201 +1,103 @@
 
 import (
-	"bytes"
 	"context"
 	"fmt"
-	"io"
-	"log"
 	"math/rand"
 	"net"
-	"net/http"
-	"os"
-	"strconv"
+	"net/url"
 	"time"
 
 	"github.com/docker/docker/api/types"
 	"github.com/docker/docker/api/types/filters"
-	"github.com/docker/docker/client"
-	"github.com/gorilla/mux"
-
-	"github.com/openfaas/faas/gateway/requests"
+	"github.com/docker/docker/api/types/swarm"
 )
 
 const watchdogPort = 8080
+const urlScheme = "http"
 
-//FunctionProxy passes-through to functions
-func FunctionProxy(wildcard bool, client *client.Client) http.HandlerFunc {
-
-	proxyClient := http.Client{
-		Transport: &http.Transport{
-			Proxy: http.ProxyFromEnvironment,
-			DialContext: (&net.Dialer{
-				Timeout:   3 * time.Second,
-				KeepAlive: 0,
-			}).DialContext,
-			MaxIdleConns:          1,
-			DisableKeepAlives:     true,
-			IdleConnTimeout:       120 * time.Millisecond,
-			ExpectContinueTimeout: 1500 * time.Millisecond,
-		},
-	}
-
-	return func(w http.ResponseWriter, r *http.Request) {
-		if r.Body != nil {
-			defer r.Body.Close()
-		}
-
-		switch r.Method {
-		case http.MethodPost,
-			http.MethodPut,
-			http.MethodPatch,
-			http.MethodDelete,
-			http.MethodGet:
-
-			xFunctionHeader := r.Header["X-Function"]
-			if len(xFunctionHeader) > 0 {
-				log.Print("X-Function: ", xFunctionHeader)
-			}
-
-			// getServiceName
-			var serviceName string
-			if wildcard {
-				muxVars := mux.Vars(r)
-
-				name := muxVars["name"]
-				serviceName = name
-			} else if len(xFunctionHeader) > 0 {
-				serviceName = xFunctionHeader[0]
-			}
-
-			if len(serviceName) > 0 {
-				lookupInvoke(w, r, serviceName, client, &proxyClient)
-			} else {
-				w.WriteHeader(http.StatusBadRequest)
-				w.Write([]byte("Provide an x-function header or valid route /function/function_name."))
-			}
-			break
-		default:
-			w.WriteHeader(http.StatusMethodNotAllowed)
-		}
-	}
+// ServiceLister is the subset of the Docker client.ServiceAPIClient needed to enable the
+// function lookup
+type ServiceLister interface {
+	ServiceList(context.Context, types.ServiceListOptions) ([]swarm.Service, error)
 }
 
-func lookupInvoke(w http.ResponseWriter, r *http.Request, name string, c *client.Client, proxyClient *http.Client) {
-	exists, err := lookupSwarmService(name, c)
-
-	if err != nil || exists == false {
-		if err != nil {
-			log.Printf("Could not resolve service: %s error: %s.", name, err)
-		}
-
-		// TODO: Should record the 404/not found error in Prometheus.
-		writeHead(name, http.StatusNotFound, w)
-		w.Write([]byte(fmt.Sprintf("Cannot find service: %s.", name)))
-	}
-
-	if exists {
-		forwardReq := requests.NewForwardRequest(r.Method, *r.URL)
-
-		invokeService(w, r, name, forwardReq, proxyClient)
-	}
+// FunctionLookup is a openfaas-provider proxy.BaseURLResolver that allows the
+// caller to verify that a function is resolvable.
+type FunctionLookup struct {
+	docker ServiceLister
+	// dnsrr controls  if DNSRoundRobin is used to resolve function
+	dnsrr bool
```

Comment:
I think `dnsrr` should also be `DNSRoundRobin` here like we changed it in the config struct - so `dnsRoundRobin` for the private var.